group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -1466,17 +1477,25 @@ public <T> CompletableFuture<T> scheduleTransactionalWriteOperation(
     ) {
         throwIfNotRunning();
         log.debug("Scheduled execution of transactional write operation {}.", name);
-        CoordinatorWriteEvent<T> event = new CoordinatorWriteEvent<>(
-            name,
+        return partitionWriter.maybeStartTransactionVerification(
             tp,
             transactionalId,
             producerId,
-            producerEpoch,
-            timeout,
-            op
-        );
-        enqueue(event);
-        return event.future;
+            producerEpoch
+        ).thenCompose(verificationGuard -> {
```

Comment:
I'm glad this ended up being kind of elegant :)

Reply:
yeah, i am also happy with this. it is simple and elegant.